### ISSUE TITLE
(aws-cdk): Can't use CLI credentials when deploying stacks in CDK v2, causing issues with default EKS RBAC

### ISSUE THREAD

**Issue by mipearson** [aws/aws-cdk#16888]
### General Issue

Can't use CLI credentials when deploying stacks in CDK v2, causing issues with default EKS RBAC

### The Question

There doesn't seem to be a way to fall back to the old behaviour of using the CLI credentials to deploy CF stacks rather than the execution role that CDK Boostrap creates.

This is an issue for EKS, where new clusters are created by default with the creator's role being the only one with master permissions. If you're using `CfnCluster` rather than the custom resource based provider, this means that you can't then fix the RBAC after the stack has finished creating, as the only entity that the cluster "trusts" is the cloudformation-only deploy role.

I would prefer not to use the CDK custom resource based provider.

### CDK CLI Version

2.0.0-rc.23 (build 1e54fb9)

### Language

Typescript

**Comment by stevehodgkiss** [comment]
I ran into this issue upgrading to CDK 2 today. I realised due a `PutKeyPolicy` failure, which was due to the cfn-exec role not being on the key's admin list (whereas the role provided via aws env vars credentials is).

There are other scenarios where we'd like to be using the role provided via env credentials, and not a cfn-exec role with extremely wide permissions. For example, we typically run CFN deploys with roles that don't allow data destruction (to prevent accidental RDS "replacements" etc). To perform those actions a separate role needs to be assumed in order to make the action intentional (i.e `aws-vault exec my-account-data-destroyer -- yarn cdk deploy ...`). It's not possible to use roles like this anymore if `cdk deploy` can't be configured to use credentials from the environment.

It would be great if the use of an assumed `cloudFormationExecutionRole` could be disabled somehow (i.e. via `cdk.json`), so that deploys behave similar to CDK 1 and use the provided credentials in ENV when available.

This is a workaround for `DefaultStackSynthesizer` that avoids having to replace the `synthesize` function:

```typescript
class DefaultStackSynthesizerWithoutDefaultCloudFormationExecutionRole extends DefaultStackSynthesizer {
  protected emitStackArtifact(stack: Stack, session: ISynthesisSession, options: SynthesizeStackArtifactOptions = {}) {
    super.emitStackArtifact(stack, session, {
      ...options,
      cloudFormationExecutionRoleArn: undefined,
    });
  }
};

// Then make sure it's used in every cdk.Stack:
export class MyBaseCdkStack extends cdk.Stack {
  constructor(scope: Construct, id: string, props: cdk.StackProps) {
    super(scope, id, {
      ...props,
      synthesizer: new DefaultStackSynthesizerWithoutDefaultCloudFormationExecutionRole(),
    });
  }
}
```

Using the `LegacyStackSynthesizer` is also an option. I added `"@aws-cdk/core:newStyleStackSynthesis": false` to `cdk.json`, and it requires the new CDKToolkit stack from CDK v2 bootstrap, and works as expected using the provided role credentials in ENV.

**Comment by mipearson** [comment]
> Thank you for posting the problem and solution here, this will be helpful for anyone else who runs into this issue. Do you think this is worth a bug report @mipearson?

I thought that this was a bug report, or at least an issue with the code documentation in BootstraplessStackSynthesizer

**Comment by peterwoodworth** [comment]
This was submitted as a guidance issue - I'll update the labels

**Comment by mipearson** [comment]
Yep - I wasn't really sure what *type* of issue it was - it's either a docs issue (working as intended, docs wrong) or a bug. I'm hoping for the latter as the examples above give situations where this functionality is needed.

**Comment by rix0rrr** [comment]
I see how `BootstraplessStackSynthesizer` is a bit of a misnomer. Sorry! It should actually be interpreted as `StagingBucketlessStackSynthesizer`, but you're right that it DOES require the roles from SOME region.

Your options are:

* Legacy synthesizer
* Your own synthesizer
* A custom resource that will confer the required permissions to some other role than the caller's role.

**Comment by samjarrett** [comment]
This change isn't just affecting eks, it impacts a bunch of permission problems and isn't documented anywhere, and potentially violates principal-of-least-privilege by mandatory escalating privs instead of leveraging permissions of the caller.  

I'm unsure why v2 deploy cli has this as an optional argument but it's basically ignored and leverages the bootstrap one and no way of disabling the behaviour

**Comment by mipearson** [comment]
@samjarrett is correct.

There's three issues here:

(a) The first is that the documentation of `BootstraplessStackSynthesizer` is incorrect: it states that it'll use the CLI credentials, but does not, as there's a `||=` in the code of the class `BootstraplessStackSynthesizer` is extending that reverts to using the created role as a default.

(b) Creating an EKS cluster via `CfnCluster` using this created role means the cluster can't be accessed, as there's no way to set up the master role mapping. There may be other similar issues with other resources & people's expectations that the CLI role/user is the one used.

(c) This decision overall has some security implications that may be surprising, as @samjarrett pointed out. They were surprising to me! On the other hand, it's a nice workaround to "oops my assumed role timed out in the middle of my big stack apply".

**Comment by otaviomacedo** [comment]
@mipearson about item b), specifically, are you referring to the admin role, created here?

https://github.com/aws/aws-cdk/blob/7bbd10deb322daf8ef1504ceb84ad3c895f291ae/packages/@aws-cdk/aws-eks/lib/cluster-resource.ts#L121-L123

Or is it the masters role, created here?

https://github.com/aws/aws-cdk/blob/1faf31d1ec7ffec4c6323a050126b0b054094c63/packages/@aws-cdk/aws-eks/lib/cluster.ts#L1430-L1432

Or something else?
